File: PHP/DataStructures/LinkedList.php
<?php

class LinkedList {

    private $root;

    public function __construct() {
        $this->root = null;
    }

    public function isEmpty() {
        return $this->root == null;
    }

    // Does the list contain the value?
    public function contains($value) {
        if($this->isEmpty()) {
            return false;
        }
        $tmp = $this->root;
        while($tmp) {
            if($tmp->value == $value) {
                return true;
            }
            $tmp = $tmp->next;
        }
        return false;
    }

    // Get a value at the specified index.
    public function get($index) {
        if($this->isEmpty()) {
            throw new OutOfBoundsException("Index out of range: " . $index);
        }
        $tmp = $this->root;
        for($i = 0; $i < $index; $i++) {
            $tmp = $tmp->next;
            if(!$tmp) {
                throw new OutOfBoundsException("Index out of range: " . $index);
            }
        }
        return $tmp->value;
    }

    // Retrieve the first element in the list.
    public function getFirst() {
        if($this->root) {
            return $this->root->value;
        }
        return null;
    }

    // Get the last element in the list.
    public function getLast() {
        if($this->isEmpty()) {
            return null;
        }
        else {
            $tmp = $this->root;
            while($tmp->next) {
                $tmp = $tmp->next;
            }
            return $tmp->value;
        }
    }

    // Insert into the end of the list.
    public function add($value) {
        if($this->isEmpty()) {
            $this->root = new Node($value, null);
        }
        else {
            while($tmp->next) {
                $tmp = $tmp->next;
            }
            $tmp->next = new Node($value, null);
        }
    }

    // Insert to the beginning of the list.
    public function addFirst($value) {
        $firstNode = new Node($value, $this->root);
        $this->root = $firstNode;
    }

    // Insert after a specified value.
    public function addAfter($key, $toAdd) {
        if($this->isEmpty()) {
            throw new Exception("Specified key " . $key . " does not exist.");
        }
        $tmp = $this->root;
        while($tmp->next) {
            if($tmp->value == $key) {
                $newNode = new Node($toAdd, $tmp->next);
                $tmp->next = $newNode;
                return;
            }
            $tmp->tmp->next;
        }

        throw new Exception("Specified key " . $key . " does not exist.");
    }

    // Insert before a given value.
    public function addBefore($key, $toAdd) {
        if($this->isEmpty()) {
            throw new Exception("Specified key " . $key . " does not exist.");
        }
        $tmpPrev = $this->root;
        $tmp = $this->root->next;
        if($tmpPrev->value == $key) {
            $newNode = new Node($value, $this->root);
            $this->root = $firstNode;
        }
        while($tmp) {
            if($tmp->value == $key) {
                $newNode = new Node($toAdd, $tmp);
                $tmpPrev->next = $newNode;
            }
            $tmpPrev = $tmp;
            $tmp = $tmp->next;
        }

        throw new Exception("Specified key " . $key . " does not exist.");
    }

    // Removes a value from the list.
    public function remove($value) {
        if($this->isEmpty()) {
            return;
        }
        $tmp = $this->root;
        if(!$tmp->next && $tmp->value == $value) {
            $this->root = null;
            return;
        }
        while($tmp->next) {
            if($tmp->next->value == $value) {
                $tmp->next = $tmp->next->next;
                return;
            }
            $tmp = $tmp->next;
        }
    }

    // Removes the first node from the list if one exists.
    public function removeFirst() {
        if($this->isEmpty()) {
            return;
        }
        $this->root = $this->root->next;
    }

    // Removes the last node from the list.
    public function removeLast() {
        if($this->isEmpty()) {
            return;
        }
        $tmp = $this->root;
        if(!$tmp->next) {
            $this->root = null;
            return;
        }

        while($tmp->next->next) {
            $tmp = $tmp->next;
        }
        $tmp->next = null;
    }

    // Gets the length of the list in O(n) time. 
    // Keeping a size variable for each insert/delete would make this O(1) time.
    public function length() {
        $count = 0;
        $tmp = $this->root;
        while($tmp) {
            $count++;
            $tmp = $tmp->next;
        }
        return $count;
    }

    // Gets the index of a value, returns -1 if DNE.
    public function indexOf($value) {
        $count = 0;
        if($this->isEmpty()) {
            return -1;
        }
        $tmp = $this->root;
        while($tmp) {
            if($tmp->value == $value) {
                return $count;
            }
            $count++;
            $tmp = $tmp->next;
        }
        return -1;
    }

    // Resets a list.
    public function clear() {
        $this->root = null;
    }

    // Converts the list to a string format.
    public function __toString() {
        $res = "{";
        $tmp = $this->root;
        while($tmp) {
            $res .= (string) $tmp->value;
            if($tmp->next) {
                $res .= ", ";
            }
            $tmp = $tmp->next;
        }
        return $res;
    }
}

// Supporting Node class for the linked list
class Node {
    
    public $value;
    public $next;

    public function __construct($value, $next) {
        $this->value = $value;
        $this->next = $next;
    }
}

File: CPlusPlus/DataStructures/LinkedList.cpp
#include <stdio.h>
#include <stdexcept>

class Node {
    public:
        Node(int value)
            : value(value), next(nullptr)
        {}
        Node(int value, Node* next)
            : value(value), next(next)
        {}
        int value;
        Node* next;
};

class LinkedList {
    public:
        LinkedList()
            :size(0), root(nullptr)
        {};
        ~LinkedList();
        bool isEmpty();
        bool contains(int value);
        int get(int index);
        int getFirst();
        int getLast();
        void add(int value);
        void addFirst(int value);
        void addAfter(int key, int toAdd);
        void addBefore(int key, int toAdd);
        void remove(int value);
        int length();
    private:
        Node* root;
        int size;
};

LinkedList::~LinkedList() {
    Node* tmp = this->root;
    while(tmp != nullptr) {
        Node* deadPtr = tmp;
        tmp = tmp->next;
        delete deadPtr;
    }
    tmp = nullptr;
}


bool LinkedList::isEmpty() {
    return size == 0;
}

bool LinkedList::contains(int value) {
    if(this->isEmpty()) {
        return false;
    }
    Node* tmp = this->root;
    while(tmp != nullptr) {
        if(tmp->value == value) {
            return true;
        }
        tmp = tmp->next;
    }
    return false;
}

int LinkedList::get(int index) {
    if(this->isEmpty()) {
        throw std::out_of_range("LinkedList::get(int index) : list is empty"); 
    }
    Node* tmp = this->root;
    for(int i = 0; i < index; ++i) {
        tmp = tmp->next;
        if(tmp == nullptr) {
            throw std::out_of_range("LinkedList::get(int index) : index out of range");
        }
    }
    return tmp->value;
}

int LinkedList::getFirst() {
    if(this->root == nullptr) {
        throw std::out_of_range("LinkedList::getFirst() : list is empty");
    }
    return this->root->value;
}

int LinkedList::getLast() {
    if(this->root == nullptr) {
        throw std::out_of_range("LinkedList::getLast() : list is empty");
    }
    Node* tmp = this->root;
    while(tmp->next != nullptr) {
        tmp = tmp->next;
    }
    return tmp->value;
}

void LinkedList::add(int value) {
    if(this->isEmpty()) {
        this->root = new Node(value);
        ++this->size;
    }
    Node* tmp = this->root;
    while(tmp->next != nullptr) {
        tmp = tmp->next;
    }
    tmp->next = new Node(value);
    ++this->size;
}

void LinkedList::addFirst(int value) {
    Node* tmp = new Node(value, this->root);
    this->root = tmp;
    ++this->size;
}

void LinkedList::addAfter(int key, int toAdd) {
    if(this->isEmpty()) {
        throw std::out_of_range("LinkedList::addAfter(int key, int toAdd) : list is empty"); 
    }
    Node* tmp = this->root;
    while(tmp->next != nullptr) {
        if(tmp->value == key) {
            Node* newNode = new Node(toAdd, tmp->next);
            tmp->next = newNode;
            ++this->size;
            return;
        }

        tmp = tmp->next;
    }
    // Key DNE
    throw std::exception();
}

void LinkedList::addBefore(int key, int toAdd) {
    if(this->isEmpty()) {
        throw std::out_of_range("LinkedList::addBefore(int key, int toAdd) : list is empty"); 
    }
    if(this->root->value == key) {
        Node* tmp = new Node(toAdd, this->root);
        this->root = tmp;
        ++this->size;
        return;
    } 

    Node* tmpPrev = this->root;
    Node* tmp = this->root->next;

    while(tmp != nullptr) {
        if(tmp->value == key) {
            tmpPrev->next = new Node(toAdd, tmp);
            ++this->size;
            return;
        }
        tmpPrev = tmp;
        tmp = tmp->next;
    }
    // Key DNE
    throw std::exception();
}

void LinkedList::remove(int value) {
    if(isEmpty()) {
        return;
    }

    Node* tmp = root;

    if(tmp->next == nullptr && tmp->value == value) {
        delete this->root;
        this->root = nullptr;
        this->size = 0;
    }

    while(tmp->next != nullptr) {
        if(tmp->next->value == value) {
            Node* deadPtr = tmp->next;
            tmp->next = tmp->next->next;
            delete deadPtr;
            return;
        }

        tmp = tmp->next;
    }
}

int LinkedList::length() {
    return this->size;
}